PrestaShop/PrestaShop
# Fix wrong flag on some languages

| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Some languages had the wrong flag. Read below to learn more.
| Type?         | bug fix
| Category?     | CO
| BC breaks?    | no
| Deprecations? | yes, protected method `Language::_copyNoneFlag()`
| Fixed ticket? | Fixes #16029
| How to test?  | Read below.

## Details

Up to now, language flag images were being chosen using the language's 2-letter ISO code. This worked fine for some combinations of language/countries where both the language ISO code and the country ISO code happen to be the same, like Spanish/Spain (es/es), French/France (fr/fr) and German/Germany (de/de). But this doesn't work well for some regionalized languages like Spanish/Argentina (es/ar), and is particularly bad for the reasons explained in [this comment](https://github.com/PrestaShop/PrestaShop/issues/16029#issuecomment-551079194).

Moreover, flags were being downloaded from prestashop.com every time.

This fix applies the following strategy:
- All flags are stored locally in `img/flags/` (I used the same libre [FamFamFam flag icons](http://www.famfamfam.com/lab/icons/flags/) as the ones served by prestashop.com)
- The flag is chosen using the following rule:
	1. If there's a `flag` property for the given language in `app/Resources/all_languages.json`, then the flag with that name is used
	2. Else, if a flag for the country name in the language locale code (the `AR` in `es-AR`) exists, then that flag is used
	3. Else, a generic flag with a question mark is used.

## There's no flag for my region!

If you're a Briton and you're mad at me for not adding the Brittany flag, this is solely because it wasn't included in the famfamfam pack. You can add it later by placing the image in the `admin-dev/themes/new-theme/static/flags/` directory, then edit `app/Resources/all_languages.json` to set up the `flag` parmeter to whatever your flag name is. Just keep in mind that the file must be a JPEG, and try to respect the same style as the other flags.

## Test cases

In all cases, you install the language in International > Translations > Add or update language, then visualize the list in International > Localization > Languages (yes that doesn't make much sense, eh @colinegin?)

1. Install Spanish (AR), you should see the Argentine flag (and not the Antigua and Barbuda one)
2. Install Arabic, you should see the Saudi Arabia flag
3. Install French (FR), you should see the French flag (and not the Belize one)
4. Install Catalan, you should see the Catalonian flag
5. Install Breton, you should see the French flag
6. Install French (CA), you should see the Canadian flag


---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/16334)
